From 93297ca2e4a44cab8c29ec6235a67d7086868144 Mon Sep 17 00:00:00 2001
From: Tendai Makumire <121951760+tmakumire@users.noreply.github.com>
Date: Thu, 20 Aug 2026 14:00:14 +0100
Subject: [PATCH] Add support for building the GKI (perf) kernel variant (#74)

For releases, we should ideally be using the GKI
variant as it is smaller and more performant
than the debug kernel variant. To make the kernel config
maintenance simpler for both variants, I have created a
common SBC config fragment which is used in both
variants.

By default, when doing a GKI variant kernel
build, the build system will do an GKI ABI check.
With the drivers we have enabled, the ABI check
fails. And this is because some of the drivers are
using symbols not listed in the KMI symbol list for the
v5.15 GKI kernel. To workaround this, I have
simply disabled the ABI check as we don't really
care about GKI compliance for our kernel
as this is more for Android kernels.
---
 arch/arm64/configs/kalama_le_gki.fragment     |  1 +
 .../vendor/kalama_le_consolidate.config       | 23 -------------------
 .../arm64/configs/vendor/kalama_le_sbc.config | 20 ++++++++++++++++
 build.config.msm.gki                          | 11 +++++++++
 4 files changed, 32 insertions(+), 23 deletions(-)
 create mode 100644 arch/arm64/configs/vendor/kalama_le_sbc.config

diff --git a/arch/arm64/configs/kalama_le_gki.fragment b/arch/arm64/configs/kalama_le_gki.fragment
index d75a0d1b6..661eb8a84 100644
--- a/arch/arm64/configs/kalama_le_gki.fragment
+++ b/arch/arm64/configs/kalama_le_gki.fragment
@@ -12,3 +12,4 @@ CONFIG_HWMON=y
 CONFIG_POWER_SUPPLY_HWMON=y
 CONFIG_THERMAL_HWMON=y
 # CONFIG_USB_NET_AX88179_178A is not set
+CONFIG_EXTCON=y
diff --git a/arch/arm64/configs/vendor/kalama_le_consolidate.config b/arch/arm64/configs/vendor/kalama_le_consolidate.config
index 7b8589f91..abef0839d 100644
--- a/arch/arm64/configs/vendor/kalama_le_consolidate.config
+++ b/arch/arm64/configs/vendor/kalama_le_consolidate.config
@@ -4,16 +4,13 @@ CONFIG_CMA_DEBUGFS=y
 CONFIG_CNSS2_DEBUG=y
 CONFIG_CORESIGHT_SOURCE_ETM4X=m
 CONFIG_QTI_THERMALZONE_CONFIG_DEBUG=m
-CONFIG_HWMON=y
 # CONFIG_HYP_ASSIGN_DEBUG is not set
-CONFIG_I2C_CHARDEV=m
 CONFIG_IOMMU_TLBSYNC_DEBUG=y
 CONFIG_LKDTM=m
 CONFIG_LOCALVERSION="-gki-consolidate"
 CONFIG_LOCK_TORTURE_TEST=m
 CONFIG_MHI_BUS_DEBUG=y
 CONFIG_MSM_GPI_DMA_DEBUG=y
-CONFIG_POWER_SUPPLY_HWMON=y
 # CONFIG_QCOM_DYN_MINIDUMP_STACK is not set
 CONFIG_QCOM_RTB=m
 CONFIG_QCOM_RTB_SEPARATE_CPUS=y
@@ -23,26 +20,6 @@ CONFIG_REGMAP_QTI_DEBUGFS_ALLOW_WRITE=y
 CONFIG_RPMSG_QCOM_GLINK_DEBUG=y
 CONFIG_RUNTIME_TESTING_MENU=y
 CONFIG_SCHED_WALT_DEBUG=m
-CONFIG_SENSORS_PWM_FAN=m
-CONFIG_SPI_SPIDEV=m
 CONFIG_TEST_USER_COPY=m
 CONFIG_THERMAL_HWMON=y
 CONFIG_UFS_DBG=y
-CONFIG_DEVMEM=y
-CONFIG_EXTCON=y
-CONFIG_USB_CONN_GPIO=m
-CONFIG_EXTCON_USB_GPIO=m
-CONFIG_LAN743X=m
-CONFIG_WLAN_VENDOR_REALTEK=y
-CONFIG_RTW88=m
-CONFIG_RTW88_PCI=m
-CONFIG_RTW88_8822CE=m
-CONFIG_SENSORS_SSC=m
-CONFIG_SND_SOC_TLV320AIC3X_I2C=m
-CONFIG_SND_SOC_TLV320AIC3X=m
-CONFIG_CRYPTO_MICHAEL_MIC=m
-CONFIG_WLAN_VENDOR_ATH=y
-CONFIG_ATH11K=m
-CONFIG_ATH11K_PCI=m
-CONFIG_TOUCHSCREEN_GOODIX=m
-CONFIG_REGULATOR_GPIO=m
diff --git a/arch/arm64/configs/vendor/kalama_le_sbc.config b/arch/arm64/configs/vendor/kalama_le_sbc.config
new file mode 100644
index 000000000..160037428
--- /dev/null
+++ b/arch/arm64/configs/vendor/kalama_le_sbc.config
@@ -0,0 +1,20 @@
+CONFIG_DEVMEM=y
+CONFIG_SPI_SPIDEV=m
+CONFIG_I2C_CHARDEV=m
+CONFIG_SENSORS_PWM_FAN=m
+CONFIG_USB_CONN_GPIO=m
+CONFIG_EXTCON_USB_GPIO=m
+CONFIG_LAN743X=m
+CONFIG_WLAN_VENDOR_REALTEK=y
+CONFIG_RTW88=m
+CONFIG_RTW88_PCI=m
+CONFIG_RTW88_8822CE=m
+CONFIG_SENSORS_SSC=m
+CONFIG_SND_SOC_TLV320AIC3X_I2C=m
+CONFIG_SND_SOC_TLV320AIC3X=m
+CONFIG_CRYPTO_MICHAEL_MIC=m
+CONFIG_WLAN_VENDOR_ATH=y
+CONFIG_ATH11K=m
+CONFIG_ATH11K_PCI=m
+CONFIG_TOUCHSCREEN_GOODIX=m
+CONFIG_REGULATOR_GPIO=m
diff --git a/build.config.msm.gki b/build.config.msm.gki
index 91d7ac044..4789a7b83 100644
--- a/build.config.msm.gki
+++ b/build.config.msm.gki
@@ -48,9 +48,20 @@ append_cmd POST_DEFCONFIG_CMDS 'sanity_check_abi'
 function build_defconfig_fragments() {
 	if [[ "${VARIANT}" =~ ^(gki|consolidate)$ ]]; then
 		apply_defconfig_fragment ${KERNEL_DIR}/arch/${ARCH}/configs/vendor/${MSM_ARCH}_GKI.config vendor/${MSM_ARCH}-gki_defconfig
+		apply_defconfig_fragment ${KERNEL_DIR}/arch/${ARCH}/configs/vendor/${MSM_ARCH}_sbc.config
+
+		if [ "${VARIANT}" = gki ] && [ "${MSM_ARCH}" = kalama_le ]; then
+			# This is a Yocto based distribution, so we intentionally
+			# skip the ABI check by setting the GKI build config to debug.
+			# The ABI check is only necessary if you intend to run a GKI compliant boot.img.
+			GKI_BUILD_CONFIG=common/build.config.gki-debug.aarch64
+			unset ABI_DEFINITION
+		fi
+
 		if [ "${VARIANT}" = gki ]; then
 			return
 		fi
+
 		GKI_BUILD_CONFIG=common/build.config.gki-debug.aarch64
 		GKI_GKI_BUILD_CONFIG_FRAGMENT=msm-kernel/build.config.gki_consolidate.aarch64
 		# ABI comparison isn't applicable on consolidate variant
-- 
2.41.0

